### PR TITLE
Validation for official plugins for upgrade tool

### LINKED ISSUE #972: OpenSearch Upgrade Tool Enhancements & Bug Fixes
This is a meta issue for tracking all changes related to the new `opensearch-upgrade` CLI tool.  This can also be referred to when back-porting the changes to 1.x 

- [x] Initial code for the `opensearch-upgrade` #846 
- [x]  Add validation to only install official plugins #973

### PR DESCRIPTION
**Pull request by VachaShah**
Signed-off-by: Vacha Shah <vachshah@amazon.com>

### Description
This PR adds validation for official plugins during the installation task for the upgrade tool.
 
### Issues Resolved
#972 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: distribution/tools/upgrade-cli/build.gradle [hunk 1]
```diff
@@ -14,6 +14,7 @@ archivesBaseName = 'opensearch-upgrade-cli'
 dependencies {
   compileOnly project(":server")
   compileOnly project(":libs:opensearch-cli")
+  implementation project(":distribution:tools:plugin-cli")
```
Comment (adnapibar) [inline distribution/tools/upgrade-cli/build.gradle:17]: Let's not create a dependency on `plugin-cli` - instead read from plugins.txt in `upgrade-cli`

Comment (VachaShah) [reply distribution/tools/upgrade-cli/build.gradle]: Removed the dependency on `plugin-cli`.

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java [hunk 2]
```diff
@@ -28,27 +56,44 @@ public void accept(final Tuple<TaskInput, Terminal> input) {
...
+            final Process process = processBuilder.inheritIO().start();
+            if (process.waitFor() != 0) {
+                terminal.errorPrint(Terminal.Verbosity.NORMAL, String.format(Locale.getDefault(), ERROR_MSG, plugin));
+            }
+        } catch (IOException | InterruptedException e) {
+            terminal.errorPrint(Terminal.Verbosity.NORMAL, String.format(Locale.getDefault(), ERROR_MSG, plugin) + e.getMessage());
+        }
+    }
+
+    public ProcessBuilder getCommandsBasedOnOS(String plugin, TaskInput taskInput) {
```
Comment (adnapibar) [inline distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java:86]: This returns a ProcessBuilder - the method name suggests it should return a command string.

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java [hunk 3]
```diff
@@ -28,27 +56,44 @@ public void accept(final Tuple<TaskInput, Terminal> input) {
...
+                manualPlugins.add(plugin);
             }
         }
+        if (!manualPlugins.isEmpty()) {
+            terminal.println("Please install the following custom plugins manually: " + manualPlugins);
+        }
         terminal.println("Success!" + System.lineSeparator());
     }
 
+    public void executeInstallPluginCommand(String plugin, TaskInput taskInput, Terminal terminal) {
```
Comment (adnapibar) [inline distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java:74]: Can be package private

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java (hunk 2 above)
Comment (adnapibar) [inline distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java:86]: Can be package private for testing.

**Review by adnapibar — COMMENTED** [review]
A few minor comments

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java (hunk 2 above)
Comment (VachaShah) [reply distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java]: Renamed the method.

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java (hunk 3 above)
Comment (VachaShah) [reply distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java]: Updated.

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java (hunk 2 above)
Comment (VachaShah) [reply distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java]: Updated.

**Review by adnapibar — APPROVED** [review]

**Comment by adnapibar** [comment]
start gradle check

File: distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java [hunk 4]
```diff
@@ -28,27 +56,46 @@ public void accept(final Tuple<TaskInput, Terminal> input) {
...
-        final ProcessBuilder processBuilder = new ProcessBuilder();
+        List<String> manualPlugins = new ArrayList<>();
+
         for (String plugin : taskInput.getPlugins()) {
-            // TODO - validate if plugin is an official plugin for OpenSearch.
-            final String command = taskInput.getOpenSearchBin().resolve("opensearch-plugin") + " install " + plugin;
-            if (OS.WINDOWS == OS.current()) {
-                processBuilder.command("cmd.exe", "/c", command);
+            if (OFFICIAL_PLUGINS.contains(plugin)) {
+                executeInstallPluginCommand(plugin, taskInput, terminal);
```
Comment (lukas-vlcek) [inline distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java:63]: Noob question here: What are official plugins? Who determines what can go into `plugins.txt` file? Is there any support for 3rd party plugins?
I am trying to understand how the upgrade tool can be used. For example if I manage Elasticsearch cluster 6.x with several 3rd party plugins and want to upgrade to OpenSearch cluster can I benefit from this upgrade tool?

Comment (adnapibar) [reply distribution/tools/upgrade-cli/src/main/java/org/opensearch/upgrade/InstallPluginsTask.java:63]: The term **official** rather should mean the core plugins which are part of the OpenSearch repository (under the **plugins** directory).  The plugins.txt is created as part of the distribution. There is a [working readme](https://github.com/opensearch-project/OpenSearch/blob/808d577be9cba15bbe4d10634946ae071ccbf3da/distribution/tools/upgrade-cli/README.md) on the upgrade tool.  Currently no plans to add support for 3rd party plugins.

**Pull request merged**